src/porepy/models/solution_strategy.py

```diff
@@ -508,26 +516,78 @@ def check_convergence(
-            # Enforce float to make mypy happy
-            error = float(np.linalg.norm(solution)) / np.sqrt(solution.size)
-            logger.info(f"Normalized residual norm: {error:.2e}")
-            converged = error < nl_params["nl_convergence_tol"]
+                return np.nan, np.nan, False, True
+
+            # Increment based error
+            increment_error = self.nonlinear_increment_error(solution_increment)
+            # Residual based error
+            residual_error = self.nonlinear_residual_error(residual, init_residual)
+            logger.info(f"Normalized increment error: {increment_error:.2e}")
+            logger.info(f"Normalized residual error: {residual_error:.2e}")
+            # Check convergence
+            converged_inc = increment_error < nl_params["nl_convergence_tol"]
+            # Allow for nan values in the residual error for effectively disabled
+            # convergence check.
+            converged_res = (
+                residual_error < nl_params["nl_convergence_tol_res"]
+                or nl_params["nl_convergence_tol_res"] is np.inf
+            )
+            converged = converged_inc and converged_res
             diverged = False
-            return error, converged, diverged
+
+        # Log the errors, increments and residuals
+        self.nonlinear_solver_statistics.log_error(increment_error, residual_error)
+        self.nonlinear_solver_statistics.log_increment(solution_increment)
+        self.nonlinear_solver_statistics.log_residual(residual, init_residual)
+
+        return residual_error, increment_error, converged, diverged
+
+    def nonlinear_residual_error(
+        self, residual: np.ndarray, init_residual: np.ndarray
+    ) -> float:
+        """Compute the residual error for a nonlinear iteration.
+
+        Parameters:
+            residual: Residual of current iteration.
+            init_residual: Reference residual value (initial residual expected).
+
+        NOTE: Access to initial residual allows for defining relative criteria.
+
+        Returns:
+            float: Residual error.
+
+        """
+        residual_error = np.linalg.norm(residual) / np.sqrt(residual.size)
+        return residual_error
+
+    def nonlinear_increment_error(self, solution_increment: np.ndarray) -> float:
+        """Compute the error based on the update increment for a nonlinear iteration.
+
+        Parameters:
+            solution_increment: Solution to the Newton linearization.
+
+        Returns:
+            float: Update error.
+        """
+        # Simple but fairly robust convergence criterions. More advanced options are
+        # e.g. considering errors for each variable and/or each grid separately,
+        # possibly using _l2_norm_cell
+        # We normalize by the size of the solution vector.
+        # Enforce float to make mypy happy
+        increment_error = np.linalg.norm(solution_increment) / np.sqrt(
```

Comment:
Could these comments be removed? Or possibly moved to the check_convergence() method instead?

Reply:
These comments have been inherited from the original code. I do not agree that check_convergence is a better place as check_convergence uses nonlinear_increment_error() only blindly, and it could possibly be replaced through a mixin.

The comment on mypy should be obsolete, though.